test/Test/Cardano/Prelude/GHC/Heap/NormalForm/Classy.hs

```diff
@@ -238,6 +239,51 @@ forceSeqToWhnf (SI.Seq SI.EmptyT)          k = k (SI.Seq SI.EmptyT)
 forceSeqToWhnf (SI.Seq (SI.Single a))      k = k (SI.Seq (SI.Single a))
 forceSeqToWhnf (SI.Seq (SI.Deep n l ft r)) k = k (SI.Seq (SI.Deep n l ft r))
 
+{-------------------------------------------------------------------------------
+  AllowThunksIn
+-------------------------------------------------------------------------------}
+
+data Record = Record
+  { field1 :: Int
+  , field2 :: Int
+  } deriving (Generic, Show)
+
+instance FromModel (AllowThunksIn '["field1"] Record) where
+  data Model (AllowThunksIn '["field1"] Record)
+    = RecordThunk   (Model (AllowThunksIn '["field1"] Record))
+    | RecordDefined (Model Int) (Model Int)
+
+  modelIsNF ctxt = \case
+    RecordThunk _ -> NotWHNF ctxt'
+    RecordDefined a b -> constrNF [modelIsNF ctxt' a, modelIsNF ctxt' b]
+    where
+      ctxt' = "Record" : ctxt
+
+  modelUnexpected ctxt m =
+    let ctxt' =
+          case modelIsNF ctxt m of
+            IsNF      -> Nothing
+            IsWHNF  a -> Just a
+            NotWHNF a -> Just a
+    in case m of
+      RecordThunk _ -> ctxt'
+      RecordDefined _ (IntThunk _) -> ctxt'
```

Comment:
I think this isn't quite right. You are returning the context computed by `modelIsNF` in the case that the right int is a thunk; but what if the left `Int` is also a thunk? This only works by accident: if the left field would have had a different type, then this would start to fail. So maybe that would a be a good idea (give the two fields different types).

I also find the definition of `modelUnexpected` slightly confusing, because we are doing a repeated pattern match on the same argument `m` within `modelUnexpected`, and moreover call `modelNF` on the exact same value. If we inline some stuff we end up with

```haskell

  modelUnexpected ctxt m =
    case m of
      RecordThunk _ -> Just ctxt' 
      RecordDefined x (IntThunk y) -> mustBeWHNF $ constrNF [modelIsNF ctxt' x, modelIsNF ctxt' (IntThunk y)] of
      RecordDefined x (IntValue y) -> Nothing
    where
      ctxt' = "Record" : ctxt

      mustBeWHNF :: IsNormalForm a -> a
      mustBeWHNF (IsWHNF a) = a
      mustBeWHNF _otherwise = error "mustBeWHNF: expected WHNF"
```

(this is just inlining your definitions), which also makes the bug I mentioned above clearer: we really shouldn't be looking at `x` at all in the `IntThunk` case. Getting rid of looking at `x` gives us

```haskell
  modelUnexpected ctxt m =
    case m of
      RecordThunk _ -> Just ctxt' 
      RecordDefined _ (IntThunk y) -> modelIsNF ctxt' (IntThunk y)
      RecordDefined _ (IntValue y) -> Nothing
    where
      ctxt' = "Record" : ctxt
```

but now we can see that we are duplicating some work that `modelUnexpected` for `Int` is already doing; so this can simplify to just

```haskell
  modelUnexpected ctxt m =
    case m of
      RecordThunk   _   -> Just ctxt' 
      RecordDefined _ y -> modelIsNF' ctxt y
    where
      ctxt' = "Record" : ctxt
```

which is a lot easier to understand: if the think itself is a thunk, this is _definitely_ a thunk; otherwise, we just need to look at the second field.